server/book/orderpq.go

```diff
@@ -34,6 +34,95 @@ type OrderPQ struct {
 	orders   map[string]*orderEntry
 }
 
+// Copy makes a deep copy of the OrderPQ. The OrderPricers are the same; each
+// orderEntry is new. The capacity and lessFn are the same.
+func (pq *OrderPQ) Copy() *OrderPQ {
+	pq.mtx.Lock()
+	defer pq.mtx.Unlock()
+	return pq.copy(pq.capacity)
+}
+
+// Copy makes a deep copy of the OrderPQ with the given capacity. The
+// OrderPricers are the same; each orderEntry is new. The lessFn is the same.
+func (pq *OrderPQ) Realloc(newCap uint32) {
+	pq.mtx.Lock()
+	defer pq.mtx.Unlock()
+	newPQ := pq.copy(newCap)
+	pq.capacity = newCap
+	pq.orders = newPQ.orders
+	pq.oh = newPQ.oh
+}
+
+// copy makes a deep copy of the OrderPQ. The OrderPricers are the same; each
+// orderEntry is new. The lessFn is the same. This function is not thread-safe.
+func (pq *OrderPQ) copy(newCap uint32) *OrderPQ {
+	// Deep copy the order heap.
+	orderHeap := make(orderHeap, 0, newCap)
+	for _, oe := range pq.oh {
+		orderHeap = append(orderHeap, &orderEntry{
+			oe.OrderPricer,
+			oe.heapIdx,
+		})
+	}
+
+	// Deep copy the orders map.
+	orders := make(map[string]*orderEntry, newCap)
+	for _, oe := range orderHeap {
+		orders[oe.UID()] = oe
+	}
+
+	newPQ := &OrderPQ{
+		oh:       orderHeap,
+		capacity: newCap,
+		lessFn:   pq.lessFn,
+		orders:   orders,
+	}
+
+	// Since the heap is copied in the same order, and with the same heap
+	// indexes, it should not be necessary to reheap. But do it to be safe.
+	heap.Init(newPQ)
+
+	return newPQ
+}
+
+// Orders copies all orders, sorted with the lessFn. The is heap sort. To avoid
+// modifying the OrderPQ or any of the data fields, a deep copy of the OrderPQ
+// is made and ExtractBest is called until all entries are extracted.
+func (pq *OrderPQ) Orders() []OrderPricer {
+	return pq.OrdersN(len(pq.oh))
```

Comment:
I was curious about whether repeated extraction was the right approach when grabbing the entire queue, so I ran some tests.

![image](https://user-images.githubusercontent.com/6109680/65147543-f99b0300-d9e3-11e9-9c8d-c4cb49fcd74c.png)

I ran `TestLargeOrderMaxPriorityQueue_Orders` and got 

![image](https://user-images.githubusercontent.com/6109680/65147624-294a0b00-d9e4-11e9-936c-e54e6fbd7e44.png)

Seems like when doing the entire queue, `sort.Slice` is faster. There may even be a speed up for `sort.Slice` because the heap ordering is a sort of partial ordering. I may butcher this, but I think that the the binary tree pop algorithm requires a sort of "percolation" step after every extraction, so is maybe not ideal for doing the entire slice, compared to other sorts.

Reply:
That's definitely correct regarding the partial sorting of the heap helping `sort.Slice`, as well as the point about a percolation/sift-down step that is applied when the best element is extracted (although it is log time).  And that's also a good point regarding simply copying the heap slice and sorting that.  I the ~3x slowdown of using repeated extractions is a solid case for just doing `sort.Slice`.  Will change this, but also keep a complete extraction method available for debugging and validation.

Reply:
> Is taking the `len(pq.oh)` a race condition here?

I believe this is fine.  Perhaps I misunderstand though.

Reply:
I switched to using `sort` in https://github.com/decred/dcrdex/pull/32/commits/0b5aa5e1656d7bd912e360e6cbfadbeda1fbb348, but I'm not seeing much performance difference.

Reply:
oops, I did that in the wrong spot, without regard to count

Reply:
OK, done right in https://github.com/decred/dcrdex/pull/32/commits/22e60c65cdd7693a74469f074bc868778aff7c27.  The performance difference is not too dramatic though.  With the `bigList` pre-generated, `TestLargeOrderMaxPriorityQueue_Orders` takes 3.38s to pop all the orders, while it takes 2.11s to sort a copy of the heap slice.  Still worth while, but the extraction isn't as bad as 6 sec for some reason.

Reply:
If I isolate the timing to a modified Orders like you had it:

(running sort then extraction)
BOOKTEST: 658 ms for sort
BOOKTEST: 1926 ms for extraction

(running extraction then sort)
BOOKTEST: 2005 ms for extraction
BOOKTEST: 656 ms for sort

So extraction takes roughly 3x longer than a simple sort.

Reply:
I wouldn't use `sort.Slice` for anything less than the entire queue. I imagine you'd see some improvement even at half the queue, but if I was to guess how these methods are going to be used, it'll probably be either small `N` or everything.

Reply:
Agreed.  I realized I made the changes right after I pushed.  The correct changes should be force pushed to the branch now, in https://github.com/decred/dcrdex/commit/22e60c65cdd7693a74469f074bc868778aff7c27

Reply:
> > Is taking the `len(pq.oh)` a race condition here?
> 
> I believe this is fine. Perhaps I misunderstand though.

I thought that this was the only `pq.oh` read/write without mutex protection, but I now see there's no lock for `Worst -> leafNodes` either. Is that okay?

Reply:
Ahh. Sorry, I was being dense.  Worst and Orders both need locking.  Resolved in https://github.com/decred/dcrdex/pull/32/commits/a27f50db8981da5cf3dcc84219223d8b8387e8d1.